subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -150,13 +151,34 @@ protected void compile(InputChanges inputChanges) {
         checkGroovyClasspathIsNonEmpty();
         warnIfCompileAvoidanceEnabled();
 
-        if (inputChanges != null && getOptions().isIncremental()) {
-            doIncrementalCompile(inputChanges);
+        GroovyJavaJointCompileSpec spec = createSpec();
+
+        if (inputChanges != null && spec.incrementalCompilationEnabled()) {
+            doIncrementalCompile(spec, inputChanges);
         } else {
-            doCompile(inputChanges, null);
+            doCompile(spec, inputChanges, null);
+        }
+    }
+
+    private void doIncrementalCompile(GroovyJavaJointCompileSpec spec, InputChanges inputChanges) {
+        Multimap<String, String> oldMappings = readSourceClassesMappingFile(getSourceClassesMappingFile());
+        getSourceClassesMappingFile().delete();
+
+        WorkResult result = doCompile(spec, inputChanges, oldMappings);
+
+        if (result instanceof IncrementalCompilationResult) {
```

Comment:
Can you add a test for the `RecompilationNotNecessary` case as well?

Reply:
Added. PTAL!